src/services/executionEnvironment.ts

```diff
@@ -0,0 +1,309 @@
+        );
+
+        const builtin_plugin_locations: string[] = [];
+        ansibleConfig.module_locations.forEach((modulePath) => {
+          const pluginsPathParts = modulePath.split(path.sep).slice(0, -1);
+          if (pluginsPathParts.includes('site-packages')) {
+            // ansible-config returns default builtin configured module path
+            // as ``<python-path>/site-packages/ansible/modules`` to copy other plugins
+            // to local cache strip the ``modules`` part from the path and append
+            // ``plugins`` folder.
+            pluginsPathParts.push('plugins');
+          }
+          builtin_plugin_locations.push(pluginsPathParts.join(path.sep));
+        });
+        // Copy builtin plugins
+        await this.copyPluginDocFiles(
+          hostCacheBasePath,
+          containerName,
+          builtin_plugin_locations,
+          '*'
+        );
+
+        // Copy builtin modules
+        ansibleConfig.module_locations = await this.copyPluginDocFiles(
+          hostCacheBasePath,
+          containerName,
+          ansibleConfig.module_locations,
+          '**/modules'
+        );
+      }
+    } catch (error) {
+      this.connection.window.showErrorMessage(
+        `Exception in ExecutionEnvironment service while fetching docs: ${JSON.stringify(
+          error
+        )}`
+      );
+    } finally {
+      this.cleanUpContainer(containerName);
+    }
+  }
+
+  public wrapContainerArgs(command: string): string {
+    const workspaceFolderPath = URI.parse(
+      this.context.workspaceFolder.uri
+    ).path;
+    const containerCommand: Array<string> = [this._container_engine];
+    containerCommand.push(...['run', '--rm']);
+    containerCommand.push(...['--workdir', workspaceFolderPath]);
+
+    containerCommand.push(
+      ...['-v', `${workspaceFolderPath}:${workspaceFolderPath}`]
+    );
+    if (this._container_engine === 'podman') {
+      // container namespace stuff
+      containerCommand.push('--group-add=root');
+      containerCommand.push('--ipc=host');
+
+      // docker does not support this option
+      containerCommand.push('--quiet');
+    } else {
+      containerCommand.push(`--user=${process.getuid()}`);
+    }
+    containerCommand.push(`--name ansible_language_server_${uuidv4()}`);
+    containerCommand.push(this._container_image);
+    containerCommand.push(command);
+    const generatedCommand = containerCommand.join(' ');
+    this.connection.console.log(
+      `container engine invocation: ${generatedCommand}`
+    );
+    return generatedCommand;
+  }
+
+  public cleanUpContainer(containerName: string): void {
+    [
+      `${this._container_engine} stop ${containerName}`,
+      `${this._container_engine} rm ${containerName}`,
+    ].forEach((command) => {
+      try {
+        child_process.execSync(command, {
+          cwd: URI.parse(this.context.workspaceFolder.uri).path,
+        });
+      } catch (error) {
+        this.connection.console.log(
```

Comment:
If the container is already stopped, or removed, there should be no message logged.

Reply:
fixed

Reply:
I was thinking about looking into the error message to determine whether it is about absent image or not, but this will do, since it's only preparation/cleanup.